File: mobile/lib/core/canonical.dart
/// Canonical JSON for bulletin payloads.
///
/// MUST produce byte-identical output to:
///   - api/src/crypto.rs canonical_bulletin_bytes()
///   - web/src/lib/canonical.ts canonicalBulletin()
///
/// Rules:
///   - Keys sorted alphabetically.
///   - No trailing newline.
///   - No whitespace between separators.
library;

import 'dart:convert' show jsonEncode;

dynamic _canonicalize(dynamic v) {
  if (v is Map) {
    final sorted = <String, dynamic>{};
    final keys = v.keys.cast<String>().toList()..sort();
    for (final k in keys) {
      sorted[k] = _canonicalize(v[k]);
    }
    return sorted;
  }
  if (v is List) {
    return v.map(_canonicalize).toList();
  }
  return v;
}

String canonicalBulletin(Map<String, dynamic> payload) {
  return jsonEncode(_canonicalize(payload));
}

File: web/src/lib/canonical.ts
/**
 * Canonical JSON bytes for bulletin payloads.
 *
 * MUST produce byte-identical output to:
 *   - api/src/crypto.rs canonical_bulletin_bytes()
 *   - mobile/lib/core/canonical.dart canonicalBulletin()
 *
 * Rules:
 *   - Keys are sorted alphabetically.
 *   - No trailing newline.
 *   - No whitespace between separators.
 *
 * If any of the three components drift, signature verification will fail.
 */

export interface BulletinPayload {
  kind: string;
  title: string;
  body: string;
  created_at: string;
}

function canonicalize(value: unknown): unknown {
  if (value === null || typeof value !== 'object') return value;
  if (Array.isArray(value)) return value.map(canonicalize);
  const obj = value as Record<string, unknown>;
  const sorted: Record<string, unknown> = {};
  for (const k of Object.keys(obj).sort()) {
    sorted[k] = canonicalize(obj[k]);
  }
  return sorted;
}

export function canonicalBulletin(payload: BulletinPayload): string {
  return JSON.stringify(canonicalize(payload));
}